apps/server/src/modules/deletion/domain/service/deletion-batch.service.ts

```diff
@@ -0,0 +1,151 @@
+@Injectable()
+export class DeletionBatchService {
+	constructor(
+		private readonly deletionBatchRepo: DeletionBatchRepo,
+		private readonly deletionBatchSummaryRepo: DeletionBatchSummaryRepo,
+		private readonly deletionRequestRepo: DeletionRequestRepo,
+		private readonly deletionRequestService: DeletionRequestService
+	) {}
+
+	public async createDeletionBatch(
+		params: CreateDeletionBatchParams,
+		validUserIds: EntityId[],
+		invalidUserIds: EntityId[] | undefined
+	): Promise<DeletionBatchSummary> {
+		const newBatch = new DeletionBatch({
+			id: new ObjectId().toHexString(),
+			name: params.name,
+			targetRefDomain: params.targetRefDomain,
+			targetRefIds: validUserIds,
+			createdAt: new Date(),
+			updatedAt: new Date(),
+		});
+
+		await this.deletionBatchRepo.save(newBatch);
+
+		const summary = await this.buildSummary(newBatch, invalidUserIds);
+
+		return summary;
+	}
+
+	public async getDeletionBatchDetails(batchId: EntityId): Promise<DeletionBatchDetails> {
+		const deletionBatch = await this.deletionBatchRepo.findById(batchId);
+
+		const failedDeletions: DeletionRequest[] = await this.deletionRequestRepo.findFailedByTargetRefId(
+			deletionBatch.targetRefIds
+		);
+		const failedDeletionUserIds: EntityId[] = failedDeletions.map((deletionRequest) => deletionRequest.targetRefId);
+
+		const pendingDeletions: DeletionRequest[] = await this.deletionRequestRepo.findPendingByTargetRefId(
+			deletionBatch.targetRefIds
+		);
+		const pendingDeletionUserIds: EntityId[] = pendingDeletions.map((deletionRequest) => deletionRequest.targetRefId);
+
+		const successfulDeletions: DeletionRequest[] = await this.deletionRequestRepo.findSuccessfulByTargetRefId(
+			deletionBatch.targetRefIds
+		);
+		const successfulDeletionUserIds: EntityId[] = successfulDeletions.map(
+			(deletionRequest) => deletionRequest.targetRefId
+		);
+
+		const summary: DeletionBatchDetails = {
+			id: deletionBatch.id,
+			pendingDeletions: pendingDeletionUserIds,
+			failedDeletions: failedDeletionUserIds,
+			successfulDeletions: successfulDeletionUserIds,
+		};
+
+		return summary;
+	}
+
+	public async getDeletionBatchSummaries(
+		findOptions: IFindOptions<DeletionBatchSummary>
+	): Promise<Page<DeletionBatchSummary>> {
+		const deletionBatches: Page<DeletionBatch> = await this.deletionBatchRepo.findDeletionBatches(findOptions);
+
+		const summaries: DeletionBatchSummary[] = await Promise.all(
+			deletionBatches.data.map(async (batch) => {
+				const summary = await this.buildSummary(batch);
+				return summary;
+			})
+		);
+
+		const page: Page<DeletionBatchSummary> = {
+			data: summaries,
+			total: deletionBatches.total,
+		};
+
+		return page;
+	}
+
+	public async requestDeletionForBatch(batch: DeletionBatch): Promise<DeletionBatchSummary> {
+		for (const targetRefId of batch.targetRefIds) {
+			const deleteNow = new Date(Date.now());
+			await this.deletionRequestService.createDeletionRequest(targetRefId, DomainName.USER, deleteNow);
+		}
```

Comment:
```suggestion
		 const deleteNow = new Date();
                 const deletionRequests = batch.targetRefIds.map(targetRefId => this.deletionRequestService.createDeletionRequest(targetRefId, DomainName.USER, deleteNow)
  );
                  await Promise.all(deletionRequests);
```

Reply:
on a second thought, this Promise.all might not be a good idea
Because it can be that we try to create a deletion request for a user id which already exists, in which case we should get an exception. And promise.all would stop